pkg/egresspolicy/egresspolicy.go

```diff
@@ -0,0 +1,566 @@
+	if len(podIPs) == 0 {
+		return
+	}
+
+	var podData *podMetadata
+	var err error
+	if podData, err = manager.getPodMetadata(pod, podIPs); err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.K8sPodName:   pod.Name,
+			logfields.K8sNamespace: pod.Namespace,
+		}).Error("failed to get valid pod metadata")
+		return
+	}
+
+	if policies, ok := manager.policyPods[id]; ok {
+		for _, policy := range policies {
+			config := manager.policyConfigs[policy]
+			manager.deleteEgressMap(config, podData)
+		}
+		delete(manager.policyPods, id)
+	}
+	delete(manager.podDataStore, id)
+}
+
+func (manager *Manager) getLocalPodsForNamespace(ns *slim_corev1.Namespace) []*slim_corev1.Pod {
+	var retPods []*slim_corev1.Pod
+
+	podStore := manager.storeGetter.GetStore("pod")
+	for _, podItem := range podStore.List() {
+		pod, ok := podItem.(*slim_corev1.Pod)
+		if !ok {
+			continue
+		}
+
+		if pod.Namespace != ns.Name {
+			continue
+		} else {
+			retPods = append(retPods, pod)
+		}
+	}
+
+	return retPods
+}
+
+func (manager *Manager) OnUpdateNamespace(ns *slim_corev1.Namespace) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+	// Ns labels have been updated, need to re-evaluate all pods within ns.
+	pods := manager.getLocalPodsForNamespace(ns)
+	if len(pods) == 0 {
+		return
+	}
+	for _, pod := range pods {
+		manager.OnUpdatePodLocked(pod, true /* removeOld */, true /* upsertNew */)
+	}
+}
+
+func (manager *Manager) getPodMetadata(pod *slim_corev1.Pod, podIPs []string) (*podMetadata, error) {
+	podId := types.NamespacedName{
+		Name:      pod.GetName(),
+		Namespace: pod.GetNamespace(),
+	}
+
+	namespace, err := manager.getNamespaceForPod(pod)
+	if namespace == nil || err != nil {
+		log.Warn("Failed in getting namespace")
+		data := &podMetadata{
+			ips:    podIPs,
+			labels: pod.GetLabels(),
+			id:     podId,
+		}
+
+		// Keep id to pod data mapping up to date
+		manager.podDataStore[podId] = data
+		return data, nil
+	}
+
+	data := &podMetadata{
+		ips:      podIPs,
+		labels:   pod.GetLabels(),
+		nsLabels: namespace.GetLabels(),
+		id:       podId,
+	}
+
+	// Keep id to pod data mapping up to date
+	manager.podDataStore[podId] = data
+	return data, nil
+}
+
+func (manager *Manager) getNamespaceForPod(pod *slim_corev1.Pod) (*slim_corev1.Namespace, error) {
+	nsName := pod.GetNamespace()
+	nsStore := manager.storeGetter.GetStore("namespace")
+	nsItem, exists, err := nsStore.GetByKey(pod.GetNamespace())
+	if err != nil {
+		log.WithError(err).Warn("Failed to get namespace")
+		return nil, err
+	}
+	if !exists {
+		log.WithFields(logrus.Fields{"nsName": nsName}).Warn("Namespace does not exist")
+		return nil, nil
+	}
+	return nsItem.(*slim_corev1.Namespace), nil
+}
+
+// isValidConfig validates the given policy config.
+func (manager *Manager) isValidConfig(config EgressPolicyConfig) error {
+	for _, policyConfig := range manager.policyConfigs {
+		if policyConfig.egressIP.String() == config.egressIP.String() {
```

Comment:
Why can't there be 2 policies which use the same egress IP? The following example:
```
spec:
  egress:
     - podSelector:
        matchLabels:
          app: app1
  destinationCidrs:
    - a.b.c.0/24
  egressSourceIp:  192.168.33.11
```

```
spec:
  egress:
     - podSelector:
        matchLabels:
          app: app2
  destinationCidrs:
    - e.f.g.0/24
  egressSourceIp:  192.168.33.11
```
where 2 different apps targeting 2 different destination CIDRs use the same egress IP would not work :thinking:

Reply:
Good question. Ideally one application should own an egress IP, as something for example to whitelist on firewall. Is there a use case where several applications want to share the same egress IP? I think the answer is probably yes, too. I think I am okay both ways.Ideas?

Reply:
What's the downside for us to support multiple source pods using the same egress IP?

Reply:
It's not a matter of multiple source pods mapping for single ip, that can be done via a selector and is supported. What we are preventing is multiple rules using same ip, which may not go well with the notion of "single owner of an egress ip". There is no difference from engineering point of view.

Reply:
I guess we can always allow it if someone comes up with a use case. Easier to allow later than to break API later, so let's keep as is.